smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaRecordBatchStream.java

```diff
@@ -25,7 +40,60 @@ public void subscribe(MultiSubscriber<? super ConsumerRecords<K, V>> subscriber)
         // Enqueue ConsumerRecords by batches, max poll records is considered 1
         KafkaRecordStreamSubscription<K, V, ConsumerRecords<K, V>> subscription = new KafkaRecordStreamSubscription<>(
                 client, config, subscriber, context, 1, (cr, q) -> q.offer(cr));
+        subscriptions.add(subscription);
         subscriber.onSubscribe(subscription);
     }
 
+    void removeFromQueueRecordsFromTopicPartitions(Collection<TopicPartition> partitions) {
```

Comment:
I don't completely understand what is happening in this method. 
I think I have a cleaner alternative for this but then maybe I am missing something. 
```java
void removeFromQueueRecordsFromTopicPartitions(Collection<TopicPartition> revokedPartitions) {
    if (revokedPartitions.isEmpty()) {
        return;
    }
    for (KafkaRecordStreamSubscription<K, V, ConsumerRecords<K, V>> subscription : subscriptions) {
        subscription.rewriteQueue(cr -> {
            Map<TopicPartition, List<ConsumerRecord<K, V>>> records = new HashMap<>();
            for (TopicPartition topicPartition : cr.partitions()) {
                if (!revokedPartitions.contains(topicPartition)) {
                    records.put(topicPartition, cr.records(topicPartition));
                }
            }
            if (!records.isEmpty()) {
                // replace ConsumerRecords with the new one
                return new ConsumerRecords<>(records);
            } else {
                // remove from the queue
                return null;
            }
        });
    }
}
```

And the `rewriteQueue` is simply the logic that you wrote but in a separate method: 
```java
void rewriteQueue(Function<T, T> mapFunction) {
    synchronized (this.queue) {
        ArrayDeque<T> replacementQueue = new ArrayDeque<>();
        Iterator<T> iterator = queue.iterator();
        while (iterator.hasNext()) {
            T newRecord = mapFunction.apply(iterator.next());
            if (newRecord != null) {
                replacementQueue.offer(newRecord);
            }
            iterator.remove();
        }
        queue.addAll((Iterable<T>) replacementQueue);
    }
}
```

Tell me what do you think

Reply:
doing an empty check is nice!

The ** removeFromQueueRecordsFromTopicPartitions** for **KafkaRecordBatchStream** is much cleaner to read. I was worried it would reorder the records in the batch but the **KafkaConsumer** does in fact fill up the batch from one topic partition before moving to the next. Regardless, we _don't have to_ guarantee order between topic partitions.

The ** rewriteQueue** method you wrote for **KafkaRecordBatchStream** works fine because we _have to_ replace all items since **ConsumerRecords** is immutable and we are talking about at most 3 items (with defaults). But for **KafkaRecordStream** it's inneficient since we only have to remove revoked records. We don't have to recreate the queue and in this case we are talking about at most 3000 item (with defaults).

I'll make some changes.

Reply:
@ozangunalp changes pushed

Reply:
Actually wrong! Performing a clear and then re appending the elements is no slower but possibly faster than doing a bulk remove.

Reply:
Done

Reply:
For `KafkaRecordStream` (`RecordQueue` with ConsumerRecord) I actually thought about using `removeIf` method inside ArrayDeque, which has a bulkRemove implementation. But what you've done is fine by me.